File: app/src/main/java/com/sortedqueue/programmercreek/fragments/ModuleDetailsFragment.java
package com.sortedqueue.programmercreek.fragments;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import com.sortedqueue.programmercreek.R;
import com.sortedqueue.programmercreek.activity.SyntaxLearnActivity;
import com.sortedqueue.programmercreek.adapter.SyntaxPagerAdapter;
import com.sortedqueue.programmercreek.database.LanguageModule;
import com.sortedqueue.programmercreek.database.SyntaxModule;
import com.sortedqueue.programmercreek.interfaces.ModuleDetailsScrollPageListener;
import com.sortedqueue.programmercreek.interfaces.SyntaxNavigationListener;
import com.sortedqueue.programmercreek.view.ScrollableViewPager;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Alok on 26/12/16.
 */

public class ModuleDetailsFragment extends Fragment implements ModuleDetailsScrollPageListener {

    @BindView(R.id.ProgressBar)
    ProgressBar progressBar;
    @BindView(R.id.syntaxLearnViewPager)
    ScrollableViewPager syntaxLearnViewPager;
    @BindView(R.id.viewPagerLayout)
    LinearLayout viewPagerLayout;
    @BindView(R.id.doneFAB)
    FloatingActionButton doneFAB;
    private LanguageModule module;
    private LanguageModule nextModule;
    private ArrayList<SyntaxModule> syntaxModules;
    private SyntaxPagerAdapter syntaxPagerAdapter;
    private SyntaxNavigationListener syntaxNavigationListener;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_module_details, container, false);
        ButterKnife.bind(this, view);
        setupViews();
        return view;
    }

    private void setupViews() {

        syntaxLearnViewPager.setOffscreenPageLimit(syntaxModules.size());
        syntaxPagerAdapter = new SyntaxPagerAdapter(getChildFragmentManager(), module, nextModule, syntaxModules, this);
        syntaxLearnViewPager.setAdapter(syntaxPagerAdapter);
        syntaxLearnViewPager.setCanScroll(true);
        progressBar.setMax(syntaxModules.size());
        progressBar.setProgress(1);
        doneFAB.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onScrollForward();
            }
        });
        toggleFabDrawable( progressBar.getProgress() );
        syntaxLearnViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                progressBar.setProgress(position + 1);
                toggleFabDrawable(progressBar.getProgress());
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

    }

    private void toggleFabDrawable(final int progress) {
        int drawable = progress == progressBar.getMax() ? R.drawable.ic_done_all : android.R.drawable.ic_media_play;
        doneFAB.setImageDrawable(ContextCompat.getDrawable(getContext(), drawable));
        syntaxNavigationListener.setImageDrawable(drawable);
    }

    public void setParameters(LanguageModule module, ArrayList<SyntaxModule> syntaxModules, LanguageModule nextModule) {
        this.module = module;
        this.syntaxModules = syntaxModules;
        this.nextModule = nextModule;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

    }

    @Override
    public void onScrollForward() {
        if (syntaxLearnViewPager.getCurrentItem() + 1 == syntaxPagerAdapter.getCount()) {
            getActivity().onBackPressed();
        } else {
            syntaxLearnViewPager.setCurrentItem(syntaxLearnViewPager.getCurrentItem() + 1);
        }

    }

    @Override
    public void toggleFABDrawable() {
        toggleFabDrawable(progressBar.getProgress() - 1);
    }

    public void setSyntaxNavigationListener(SyntaxNavigationListener syntaxNavigationListener) {
        this.syntaxNavigationListener = syntaxNavigationListener;
    }
}

File: app/src/main/java/com/sortedqueue/programmercreek/fragments/ModuleDetailsFragment.kt
package com.sortedqueue.programmercreek.fragments

import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ProgressBar

import com.sortedqueue.programmercreek.R
import com.sortedqueue.programmercreek.activity.SyntaxLearnActivity
import com.sortedqueue.programmercreek.adapter.SyntaxPagerAdapter
import com.sortedqueue.programmercreek.database.LanguageModule
import com.sortedqueue.programmercreek.database.SyntaxModule
import com.sortedqueue.programmercreek.interfaces.ModuleDetailsScrollPageListener
import com.sortedqueue.programmercreek.interfaces.SyntaxNavigationListener
import com.sortedqueue.programmercreek.view.ScrollableViewPager
import kotlinx.android.synthetic.main.fragment_module_details.*

import java.util.ArrayList




/**
 * Created by Alok on 26/12/16.
 */

class ModuleDetailsFragment : Fragment(), ModuleDetailsScrollPageListener {

    private var module: LanguageModule? = null
    private var nextModule: LanguageModule? = null
    private var syntaxModules: ArrayList<SyntaxModule>? = null
    private var syntaxPagerAdapter: SyntaxPagerAdapter? = null
    private var syntaxNavigationListener: SyntaxNavigationListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_module_details, container, false)


        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupViews()
    }

    private fun setupViews() {

        syntaxLearnViewPager!!.offscreenPageLimit = syntaxModules!!.size
        syntaxPagerAdapter = SyntaxPagerAdapter(childFragmentManager, module!!, nextModule, syntaxModules!!, this)
        syntaxLearnViewPager!!.adapter = syntaxPagerAdapter
        syntaxLearnViewPager!!.setCanScroll(true)
        ProgressBar!!.max = syntaxModules!!.size
        ProgressBar!!.progress = 1
        doneFAB!!.setOnClickListener { onScrollForward() }
        toggleFabDrawable(ProgressBar!!.progress)
        syntaxLearnViewPager!!.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                ProgressBar!!.progress = position + 1
                toggleFabDrawable(ProgressBar!!.progress)
            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })

    }

    private fun toggleFabDrawable(progress: Int) {
        val drawable = if (progress == ProgressBar!!.max) R.drawable.ic_done_all else android.R.drawable.ic_media_play
        doneFAB!!.setImageDrawable(ContextCompat.getDrawable(context!!, drawable))
        syntaxNavigationListener!!.setImageDrawable(drawable)
    }

    fun setParameters(module: LanguageModule, syntaxModules: ArrayList<SyntaxModule>, nextModule: LanguageModule?) {
        this.module = module
        this.syntaxModules = syntaxModules
        this.nextModule = nextModule
    }

    override fun onDestroyView() {
        super.onDestroyView()

    }

    override fun onScrollForward() {
        if (syntaxLearnViewPager!!.currentItem + 1 == syntaxPagerAdapter!!.count) {
            activity!!.onBackPressed()
        } else {
            syntaxLearnViewPager!!.currentItem = syntaxLearnViewPager!!.currentItem + 1
        }

    }

    override fun toggleFABDrawable() {
        toggleFabDrawable(ProgressBar!!.progress - 1)
    }

    fun setSyntaxNavigationListener(syntaxNavigationListener: SyntaxNavigationListener) {
        this.syntaxNavigationListener = syntaxNavigationListener
    }
}
